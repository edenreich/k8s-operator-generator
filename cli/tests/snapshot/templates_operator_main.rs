use askama::Template;
use insta::assert_snapshot;
use kopgen::{errors::AppError, templates::operator::Main};

#[test]
fn render_with_controllers_and_types() -> Result<(), AppError> {
    let template = Main {
        api_group: "example.com".to_string(),
        api_version: "v1".to_string(),
        controllers: vec!["controller1".to_string(), "controller2".to_string()],
        types: vec!["Type1".to_string(), "Type2".to_string()],
    };

    let rendered = template.render()?;
    assert_snapshot!(rendered);
    Ok(())
}

#[test]
fn render_with_controllers() -> Result<(), AppError> {
    let template = Main {
        api_group: "example.com".to_string(),
        api_version: "v1".to_string(),
        controllers: vec!["controller1".to_string(), "controller2".to_string()],
        types: vec![],
    };

    let rendered = template.render()?;
    assert_snapshot!(rendered);
    Ok(())
}

#[test]
fn render_with_types() -> Result<(), AppError> {
    let template = Main {
        api_group: "example.com".to_string(),
        api_version: "v1".to_string(),
        controllers: vec![],
        types: vec!["Type1".to_string(), "Type2".to_string()],
    };

    let rendered = template.render()?;
    assert_snapshot!(rendered);
    Ok(())
}

#[test]
fn render_without_controllers_or_types() -> Result<(), AppError> {
    let template = Main {
        api_group: "example.com".to_string(),
        api_version: "v1".to_string(),
        controllers: vec![],
        types: vec![],
    };

    let rendered = template.render()?;
    assert_snapshot!(rendered);
    Ok(())
}
