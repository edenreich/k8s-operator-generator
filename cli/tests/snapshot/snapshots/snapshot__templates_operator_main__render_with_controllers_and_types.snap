---
source: cli/tests/snapshot/templates_operator_main.rs
expression: rendered
---
/// This file is generated by kopgen. Do not edit manually. If you need to make adjustments add it to .openapi-generator-ignore file.
use anyhow::Context;
use clap::Parser;
use k8s_openapi::apiextensions_apiserver::pkg::apis::apiextensions::v1::CustomResourceDefinition;
use kube::{api::Api, Client as KubeClient, CustomResourceExt};
use log::{debug, error, info};
use operator::{
    cli::{Cli, Commands},
    deploy_crd, wait_for_crd
};
use warp::Filter;

use operator::controllers::{
        controller1,
        controller2,
};

use operator::types::{
        Type1::Type1,
        Type2::Type2,
};

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    env_logger::init();

    let cli = Cli::parse();

    debug!("Log level: {}", cli.verbosity);

    match cli.command {
        Some(Commands::Run { install_crds }) => {
            info!("Starting operator...");
            debug!("CRD's will be installed automatically: {}", install_crds);

            let kube_client = KubeClient::try_default().await?;
            let kube_client_api: Api<CustomResourceDefinition> = Api::all(kube_client.clone());

            if install_crds {
                info!("Deploying CRDs...");

                let crds = vec![
                    Type1::crd(),
                    Type2::crd(),
                ];

                for crd in crds {
                    deploy_crd(kube_client_api.clone(), crd).await?;
                }
            }

            let controllers_crds = vec![
                format!("controller1.example.com"),
                format!("controller2.example.com"),
            ];
            for controller_crd in controllers_crds {
                if let Err(e) = wait_for_crd(kube_client_api.clone(), &controller_crd).await {
                    error!("Error waiting for CRD {}: {}", &controller_crd, e);
                }
            }

            // Start the controller1 controller for the controller1.example.com/v1 API group
            let controller1_client = Api::namespaced(kube_client.clone(), "default");
            tokio::spawn(async {
                let _controller1_controller = controller1::handle(controller1_client).await;
            });

            // Start the controller2 controller for the controller2.example.com/v1 API group
            let controller2_client = Api::namespaced(kube_client.clone(), "default");
            tokio::spawn(async {
                let _controller2_controller = controller2::handle(controller2_client).await;
            });

            tokio::spawn(async {
                let liveness_route = warp::path!("healthz")
                    .map(|| warp::reply::with_status("OK", warp::http::StatusCode::OK));

                let readiness_route = warp::path!("readyz")
                    .map(|| warp::reply::with_status("OK", warp::http::StatusCode::OK));

                let health_routes = liveness_route.or(readiness_route);

                warp::serve(health_routes).run(([0, 0, 0, 0], 8000)).await;
            });
        }
        Some(Commands::Version) => {
            println!("Operator version: {}", env!("CARGO_PKG_VERSION"));
            return Ok(());
        }
        None => {
            error!("No command provided. Use --help for more information.");
            return Err(anyhow::anyhow!(
                "No command provided. Use --help for more information."
            ));
        }
    }

    tokio::signal::ctrl_c()
        .await
        .context("Failed to listen for Ctrl+C")?;
    info!("Termination signal received. Shutting down.");

    Ok(())
}
